### PR TITLE
Aggregate generation

### PR DESCRIPTION
**Pull request by serhii-lekariev**
This PR introduces scaffolding of empty aggregates.

In order to generate an aggregate, a Protobuf message containing a respective `(entity).kind` option value is necessary.

In order to generate aggregates, type dependencies are resolved to their respective java classes, which are later used for scaffolding.

### CODE REVIEW

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL.

File: code-generation/src/main/java/io/spine/tools/cli/generation/AggregateScaffolder.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            scaffolded.writeTo(scaffoldingDestination.toPath());
+        } catch (IOException e) {
+            throw newIllegalStateException(e, "Couldn't write aggregate %s to a file.",
+                                           declaration.simpleName());
+        }
+    }
+
+    private static TypeSpec classDeclaration(AggregateDeclaration declaration, ClassName identifier,
+                                             ClassName state, ClassName vBuilder,
+                                             MethodSpec constructor) {
```
[the remark this answers is not in the document]
Comment (dmdashenkov) [reply code-generation/src/main/java/io/spine/tools/cli/generation/AggregateScaffolder.java]: Seems like all the data is retrieved from an instance of `AggregateDeclaration`. Maybe, this is the object we're looking for.

File: code-generation/src/main/java/io/spine/tools/cli/generation/AggregateScaffolder.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        return TypeSpec.classBuilder(declaration.simpleName())
+                       .superclass(
+                               ParameterizedTypeName.get(AGGREGATE_SUPERCLASS,
+                                                         identifier, state,
+                                                         vBuilder))
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/AggregateScaffolder.java]: Please align the parameters. I'd also extract `ParameterizedTypeName` into a variable.

File: code-generation/src/main/java/io/spine/tools/cli/generation/AggregateScaffolder.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    private static MethodSpec constructor(ClassName identifier) {
+        CodeBlock superCall = CodeBlock.builder()
+                                       .add("super(id)")
+                                       .build();
+        return MethodSpec.constructorBuilder()
+                         .addJavadoc("It is suggested to keep constructor private, since " +
+                                             "instances of aggregates are created" +
+                                             System.lineSeparator() +
+                                             "by their respective repositories." +
+                                             System.lineSeparator())
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/AggregateScaffolder.java]: Let's discuss the doc content.

File: code-generation/src/main/java/io/spine/tools/cli/generation/AggregateTypeParameters.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+     *         type of state
+     * @param vBuilder
+     *         type of validating builder
+     */
+    public AggregateTypeParameters(IdentifierType identifier, StateType state, VBuilderType vBuilder) {
+        this.identifier = identifier;
+        this.state = state;
+        vBuilderType = vBuilder;
+    }
+    /** Returns type of identifier. */
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/AggregateTypeParameters.java]: An empty line is missing.

File: code-generation/src/main/java/io/spine/tools/cli/generation/AggregateTypeParameters.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.tools.cli.generation;
+
+import com.google.common.base.Objects;
+import com.squareup.javapoet.ClassName;
+
+/**
+ * Type parameters of the aggregate.
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/AggregateTypeParameters.java]:
```suggestion
 * Type parameters of an {@link Aggregate} subtype.
```

File: code-generation/src/main/java/io/spine/tools/cli/generation/AggregateTypeParameters.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * <p>The type parameters are as follows:
+ * <ol>
+ *     <li>Type of aggregate ID.
+ *     <li>Type of aggregates state message.
+ *     <li>Type of aggregates validating builder.
+ * </ol>
+ *
+ * <p>More on aggregate type parameters cna be learned in javadoc
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/AggregateTypeParameters.java]: Please revisit this paragraph. There are a few mistakes here.

File: code-generation/src/main/java/io/spine/tools/cli/generation/VBuilderType.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        VBuilderType type = (VBuilderType) o;
+        return Objects.equal(vBuilderType, type.vBuilderType);
+    }
+
+    /** Returns validating builder type name held by this instance. */
+    ClassName getType() {
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/VBuilderType.java]: Please leave the methods overriding `Object` at the bottom of the class.

File: code-generation/src/main/java/io/spine/tools/cli/generation/package-info.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * This package contains classes that allow to generate code for Spine projects.
+ */
+package io.spine.tools.cli.generation;
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/package-info.java]: There are a few annotations missing. Please see the packages in `core-java`. Please also make sure to:
 - check the rest of `package-info` files in this repo;
 - configure your IDE to generate proper package descriptors (i.e. with the required annotations) by default.

File: code-generation/src/test/java/io/spine/tools/cli/generation/AggregateLookupTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    @DisplayName("correctly distinguish aggregates from non-aggregates")
+    @Test
+    void testRightAmountOfAggregates() {
+        AggregateLookup aggregateLookup = new AggregateLookup(validDescriptorSet(), aggregatePackage());
+        Set<AggregateDeclaration> aggregates = aggregateLookup.collect();
+        assertEquals(1, aggregates.size());
+        AggregateDeclaration singleAggregate = aggregates.iterator().next();
+        assertEquals(describedAggregateName(), singleAggregate.simpleName());
+        DescriptorProtos.FileDescriptorSet.newBuilder();
```
Comment (dmdashenkov) [inline code-generation/src/test/java/io/spine/tools/cli/generation/AggregateLookupTest.java]: This seems to be a trace of an experiment. Please remove.

File: code-generation/src/test/java/io/spine/tools/cli/generation/given/AggregateLookupTestEnv.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                                                                             state,
+                                                                             vBuilder);
+        AggregateDeclaration expected = new AggregateDeclaration(describedAggregateName,
+                                                                 typeParameters);
+        return expected;
+    }
+
+}
+
+
```
Comment (dmdashenkov) [inline code-generation/src/test/java/io/spine/tools/cli/generation/given/AggregateLookupTestEnv.java]: Some empty lines here.

File: code-generation/src/test/java/io/spine/tools/cli/generation/given/AggregateLookupTestEnv.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    /** Returns a file containing a valid descriptor set. */
+    public static File validDescriptorSet() {
+        return DESCRIPTOR_SET;
+    }
+
+    /** Returns a file that doesn't exist. */
+    public static File invalidDescriptorSet() {
+        String uuid = UUID.randomUUID()
+                          .toString();
```
Comment (dmdashenkov) [inline code-generation/src/test/java/io/spine/tools/cli/generation/given/AggregateLookupTestEnv.java]: Please use `Identifier.newUuid()`.

File: code-generation/src/test/java/io/spine/tools/cli/generation/given/AggregateLookupTestEnv.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                                                         "descriptors",
+                                                         "test",
+                                                         "known_types.desc")
+                                                    .toFile();
+
+    private AggregateLookupTestEnv() {
+    }
+
+    /** Returns a package that all looked up aggregates are going to have. */
+    public static String aggregatePackage(){
```
Comment (dmdashenkov) [inline code-generation/src/test/java/io/spine/tools/cli/generation/given/AggregateLookupTestEnv.java]: Please format all the files in the changeset. Also, please make sure that your IDE does that on each commit.

File: code-generation/src/test/proto/io/spine/tools/cli/generation/flower.proto [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+package spine.cli.garden;
+
+import "spine/options.proto";
+
+option (type_url_prefix) = "type.spine.io";
+option java_package = "io.spine.tools.cli.generation";
+option java_outer_classname = "FlowerProto";
+option java_multiple_files = true;
+import "io/spine/tools/cli/generation/identifiers.proto";
```
Comment (dmdashenkov) [inline code-generation/src/test/proto/io/spine/tools/cli/generation/flower.proto]: An empty line before the import is missing.

File: code-generation/src/test/proto/io/spine/tools/cli/generation/flower.proto [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+// A flower.
+//
+// Has a name.
+//
+message Flower {
```
Comment (dmdashenkov) [inline code-generation/src/test/proto/io/spine/tools/cli/generation/flower.proto]: A flower is hardly characterized solely by its _name_. Please think a bit further on this type.

File: code-generation/src/test/proto/io/spine/tools/cli/generation/garden.proto [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+import "io/spine/tools/cli/generation/identifiers.proto";
+import "io/spine/tools/cli/generation/flower.proto";
+
+// A garden.
+//
+// Has pots with flowers.
+//
+message Garden {
+
+    option (entity).kind = AGGREGATE;
```
Comment (dmdashenkov) [inline code-generation/src/test/proto/io/spine/tools/cli/generation/garden.proto]: Usually, we don't put an empty line between the declaration and options, considering options a part of the declaration, as if they were annotations on a Java class.

File: code-generation/src/test/proto/io/spine/tools/cli/generation/garden.proto [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+    // Identifier of this garden.
+    GardenId id = 1 [(required) = true];
+
+    // Flowers stored in this garden.
```
Comment (dmdashenkov) [inline code-generation/src/test/proto/io/spine/tools/cli/generation/garden.proto]: They're not merely stored. They __grow__ in the garden. When applying the domain driven approach, such differences are of huge importance.

File: code-generation/src/test/proto/io/spine/tools/cli/generation/withered_flowers.proto [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+option java_package = "io.spine.tools.cli.generation";
+option java_outer_classname = "WitheredFlowersProto";
+option java_multiple_files = true;
+import "io/spine/tools/cli/generation/flower.proto";
+
+// Flowers that have withered.
+//
+message WitheredFlowers {
+
+    option (entity).kind = PROJECTION;
```
Comment (dmdashenkov) [inline code-generation/src/test/proto/io/spine/tools/cli/generation/withered_flowers.proto]: Same here, please remove the empty line.

File: command-line/src/main/java/io/spine/tools/cli/ProjectSetup.java [hunk 18]
```diff
@@ -58,7 +58,9 @@
      */
     @Option(names = {"-n", "--name"}, required = true, description = PROJECT_NAME_OPTION_DESCRIPTION)
     private final ProjectName projectName = assignedViaReflection();
-
+    private final ImmutableSet<FileExtension> fileExtensionsToAcknowledge =
+            ImmutableSet.of(JAVA, PROTO, GRADLE);
+    private final Printer printing = new StandardOutputPrinter();
```
Comment (dmdashenkov) [inline command-line/src/main/java/io/spine/tools/cli/ProjectSetup.java]: An empty line between this declaration and the next field is missing. Also, the previous arrangement of fields seemed to be better, since the `@Option`s were grouped.

File: command-line/src/main/java/io/spine/tools/cli/ProjectSetup.java [hunk 19]
```diff
@@ -105,7 +104,7 @@ public void run() {
      * <p>To do so:
      * <ol>
      *     <li>clones the
-     *         <a href="https://github.com/SpineEventEngine/server-quickstart">Spine template project</a>
+     *     <a href="https://github.com/SpineEventEngine/server-quickstart">Spine template project</a>
```
Comment (dmdashenkov) [inline command-line/src/main/java/io/spine/tools/cli/ProjectSetup.java]: Please revert this change.

File: text-substitution/src/main/java/io/spine/tools/cli/project/PackageLocation.java [hunk 20]
```diff
@@ -34,15 +34,19 @@
...
     private static final char END_STATEMENT_SYMBOL = ';';
     private final String packageStatement;
     private final File describedLocation;
 
-    PackageLocation(CharSequence statement) {
+    public String packageStatement(){
+        return packageStatement;
+    }
+
+    public PackageLocation(CharSequence statement) {
```
Comment (dmdashenkov) [inline text-substitution/src/main/java/io/spine/tools/cli/project/PackageLocation.java]: The constructor should go before the methods.

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL.

File: code-generation/src/main/java/io/spine/tools/cli/generation/AggregateScaffolder.java [hunk 21]
```diff
@@ -108,14 +104,18 @@ private static MethodSpec constructor(ClassName identifier) {
...
-                                             System.lineSeparator())
+                         .addJavadoc(getConstructorJavadoc())
                          .addParameter(identifier, "id")
                          .addModifiers(PRIVATE)
                          .addStatement(superCall)
                          .build();
     }
+
+    private static String getConstructorJavadoc() {
+        return "Creates new instance of this aggregate." + System.lineSeparator() +
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/AggregateScaffolder.java]: Please move this line separator to the next line.

File: code-generation/src/test/java/io/spine/tools/cli/generation/AggregateLookupTest.java [hunk 22]
```diff
@@ -50,17 +48,19 @@ void testRightAmountOfAggregates() {
...
     @Test
     void testCorrectAggregateDefined(){
         AggregateLookup aggregateLookup = new AggregateLookup(validDescriptorSet(), aggregatePackage());
         Set<AggregateDeclaration> aggregates = aggregateLookup.collect();
-        AggregateDeclaration singleAggregate = aggregates.iterator().next();
+        AggregateDeclaration singleAggregate = getAggregate(aggregates);
         assertEquals(describedAggregate(), singleAggregate);
     }
+
+    private static AggregateDeclaration getAggregate(Set<AggregateDeclaration> aggregates) {
```
Comment (dmdashenkov) [inline code-generation/src/test/java/io/spine/tools/cli/generation/AggregateLookupTest.java]: `getSingleItem`.

File: code-generation/src/test/java/io/spine/tools/cli/generation/given/AggregateLookupTestEnv.java [hunk 23]
```diff
@@ -59,18 +58,12 @@ public static File validDescriptorSet() {
...
         return invalidFile;
     }
 
-    /** Returns a name of the aggregate described by the valid descriptor set. */
-    public static String describedAggregateName() {
-        return AGGREGATE_IN_DESCRIPTOR_SET;
-    }
-
-    /**Returns a declaration of the aggregate described by the valid descriptor set. */
+    /** Returns a declaration of the aggregate described by the valid descriptor set. */
```
Comment (dmdashenkov) [inline code-generation/src/test/java/io/spine/tools/cli/generation/given/AggregateLookupTestEnv.java]: Please add a link to the `valid` descriptor set declaration.

File: code-generation/src/test/proto/io/spine/tools/cli/generation/flower.proto [hunk 24]
```diff
@@ -28,14 +28,29 @@ option (type_url_prefix) = "type.spine.io";
...
-// Has a name.
+// Has a kind, a name and a color.
 //
 message Flower {
 
     // Name of the flower (e.g. Zachary).
     string name = 1 [(required) = true];
+
+    // A kind of a lower (e.g. Daisy).
+    FlowerKind kind = 2 [(required) = true];
```
Comment (dmdashenkov) [inline code-generation/src/test/proto/io/spine/tools/cli/generation/flower.proto]: As discussed, let's remove the duplication in `name` and `kind`.

File: code-generation/src/test/proto/io/spine/tools/cli/generation/flower.proto [hunk 25]
```diff
@@ -28,14 +28,29 @@ option (type_url_prefix) = "type.spine.io";
...
+    int32 color = 3 [(required) = true];
+
+    enum FlowerKind {
+        KINDLESS = 0;
+        LILY = 1;
+        ROSE = 2;
+        DAISY = 3;
+    }
 }
+
```
Comment (dmdashenkov) [inline code-generation/src/test/proto/io/spine/tools/cli/generation/flower.proto]: This line is redundant.

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL.

**Comment by serhii-lekariev** [comment]
@dmdashenkov, PTAL.

File: code-generation/src/main/java/io/spine/tools/cli/generation/AggregateLookup.java [hunk 26]
```diff
@@ -55,6 +57,23 @@
         this.knownTypes = knownTypes;
```
Comment (dmdashenkov) [inline code-generation/src/main/java/io/spine/tools/cli/generation/AggregateLookup.java]: Please make the constructor `private`.
